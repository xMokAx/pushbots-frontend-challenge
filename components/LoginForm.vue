<template>
  <v-form
    ref="form"
    v-model="valid"
    lazy-validation
    class="d-flex flex-column"
    @submit.prevent="onSubmit"
  >
    <v-text-field
      v-model="email"
      :rules="emailRules"
      label="E-mail"
      name="email"
      required
    ></v-text-field>

    <v-text-field
      v-model="password"
      :rules="passwordRules"
      label="Password"
      name="password"
      type="password"
      required
    ></v-text-field>

    <nuxt-link to="/register" class="body-2 font-weight-bold"
      >Don't have an account?</nuxt-link
    >
    <nuxt-link to="/reset-pw" class="body-2 font-weight-bold"
      >Forgot your password?</nuxt-link
    >

    <v-btn
      :disabled="!valid || !email || !password"
      type="submit"
      color="accent"
      block
      class="my-4"
    >
      LOGIN
    </v-btn>
  </v-form>
</template>

<script>
import { mapMutations } from 'vuex'
import { SHOW_TOAST } from '../store/toast/mutation-types'
export default {
  props: {
    loading: {
      type: Boolean,
      required: true
    }
  },
  data: () => ({
    valid: false,
    email: '',
    password: '',
    emailRules: [
      (v) => !!v || 'E-mail is required',
      (v) => /.+@.+\..+/.test(v) || 'E-mail must be valid'
    ],
    passwordRules: [
      (v) => !!v || 'Password is required',
      (v) => v.length > 4 || 'Password must be more than 4 characters'
    ]
  }),
  computed: {
    isLoggingIn: {
      get() {
        return this.loading
      },
      set() {
        this.$emit('toggle-loading')
      }
    }
  },
  methods: {
    onSubmit() {
      this.isLoggingIn = true
      this.$auth
        .loginWith('local', {
          data: {
            email: this.email,
            password: this.password
          }
        })
        .then(() => {
          this[SHOW_TOAST]({
            color: 'success',
            text: `Welcome back ${this.$auth.user.name}`
          })
        })
        .catch((e) => {
          this[SHOW_TOAST]({
            color: 'error',
            text: 'Email or password is wrong'
          })
          this.isLoggingIn = false
        })
    },
    ...mapMutations('toast', [SHOW_TOAST])
  }
}
</script>
