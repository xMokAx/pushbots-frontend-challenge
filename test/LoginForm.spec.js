import Vue from 'vue'
import Vuetify from 'vuetify'
import { mount } from '@vue/test-utils'
import LoginForm from '../components/LoginForm'

Vue.use(Vuetify)

const propsData = {
  loading: false
}

describe('LoginForm', () => {
  let wrapper
  beforeEach(() => {
    wrapper = mount(LoginForm, {
      propsData,
      stubs: ['nuxt-link'],
      sync: false
    })
  })

  it('Should match snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('Should call onSubmit on click', () => {
    const onSubmit = jest.fn()
    wrapper.setMethods({ onSubmit })
    const loginButton = wrapper.find('[type="submit"]')
    loginButton.element.disabled = false
    loginButton.trigger('submit')
    expect(onSubmit).toBeCalled()
  })
})
