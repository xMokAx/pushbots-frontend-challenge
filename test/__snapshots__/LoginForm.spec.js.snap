// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm Should match snapshot 1`] = `
"<form novalidate=\\"novalidate\\" class=\\"v-form d-flex flex-column\\"><div class=\\"v-input theme--light v-text-field\\"><div class=\\"v-input__control\\"><div class=\\"v-input__slot\\"><div class=\\"v-text-field__slot\\"><label for=\\"input-3\\" class=\\"v-label theme--light\\" style=\\"right: 0px; position: absolute;\\">E-mail</label><input name=\\"email\\" required=\\"required\\" id=\\"input-3\\" type=\\"text\\"></div></div><div class=\\"v-text-field__details\\"><div class=\\"v-messages theme--light\\"><span name=\\"message-transition\\" tag=\\"div\\" class=\\"v-messages__wrapper\\"></span></div></div></div></div> <div class=\\"v-input theme--light v-text-field\\"><div class=\\"v-input__control\\"><div class=\\"v-input__slot\\"><div class=\\"v-text-field__slot\\"><label for=\\"input-6\\" class=\\"v-label theme--light\\" style=\\"right: 0px; position: absolute;\\">Password</label><input name=\\"password\\" required=\\"required\\" id=\\"input-6\\" type=\\"password\\"></div></div><div class=\\"v-text-field__details\\"><div class=\\"v-messages theme--light\\"><span name=\\"message-transition\\" tag=\\"div\\" class=\\"v-messages__wrapper\\"></span></div></div></div></div> <nuxt-link-stub to=\\"/register\\" class=\\"body-2 font-weight-bold\\">Don't have an account?</nuxt-link-stub> <nuxt-link-stub to=\\"/reset-pw\\" class=\\"body-2 font-weight-bold\\">Forgot your password?</nuxt-link-stub> <button type=\\"submit\\" disabled=\\"disabled\\" class=\\"my-4 v-btn v-btn--block v-btn--contained v-btn--disabled theme--light v-size--default\\"><span class=\\"v-btn__content\\">
    LOGIN
  </span></button></form>"
`;
